baremetal/cluster_manager_test.go

```diff
@@ -0,0 +1,290 @@
+		BMCluster     *infrav1.BareMetalCluster
+		Cluster       *clusterv1.Cluster
+		ExpectSuccess bool
+	}{
+		"Cluster, BMCluster, BMMachine amd Machine exist": {
+			Machine:       newMachine(clusterName, machineName, bareMetalMachineName),
+			BMMachine:     newBareMetalMachine(bareMetalMachineName, bmmOwnerRef, nil, nil),
+			Cluster:       newCluster(clusterName),
+			BMCluster:     newBareMetalCluster(baremetalClusterName, bmcOwnerRef, bmcSpec, nil),
+			ExpectSuccess: true,
+		},
+		"Cluster exists, BMCluster empty, BMMachine and Machine exist": {
+			Machine:       newMachine(clusterName, machineName, bareMetalMachineName),
+			BMMachine:     newBareMetalMachine(bareMetalMachineName, bmmOwnerRef, nil, nil),
+			Cluster:       newCluster(clusterName),
+			BMCluster:     &infrav1.BareMetalCluster{},
+			ExpectSuccess: false,
+		},
+		"Cluster empty, BMCluster, BMMachine and Machine exist": {
+			Machine:       newMachine(clusterName, machineName, bareMetalMachineName),
+			BMMachine:     newBareMetalMachine(bareMetalMachineName, bmmOwnerRef, nil, nil),
+			Cluster:       &clusterv1.Cluster{},
+			BMCluster:     newBareMetalCluster(baremetalClusterName, bmcOwnerRef, bmcSpec, nil),
+			ExpectSuccess: true,
+		},
+		"Cluster and BMCluster are nil": {
+			Cluster:       nil,
+			BMCluster:     nil,
+			ExpectSuccess: false,
+		},
+		"Cluster exists and BMCluster is nil": {
+			Cluster:       newCluster(clusterName),
+			BMCluster:     nil,
+			ExpectSuccess: false,
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Logf("## TC-%s ##", name)
+
+		objects := []runtime.Object{
+			tc.Cluster, tc.BMCluster, tc.BMMachine, tc.Machine,
+		}
+
+		if tc.Cluster == nil || tc.BMCluster == nil {
+			tc.c = fakeclient.NewFakeClientWithScheme(setupScheme())
+		} else {
+			tc.c = fakeclient.NewFakeClientWithScheme(setupScheme(), objects...)
+		}
+
+		clusterMgr, err := newClusterManager(tc.c, tc.Cluster, tc.BMCluster, klogr.New())
+
+		if err != nil {
+			if tc.ExpectSuccess {
+				t.Error(err)
+			}
+		}
+
+		if clusterMgr == nil {
+			if tc.ExpectSuccess {
+				t.Error(err)
+			}
+		} else {
+			err = clusterMgr.Create(context.TODO())
+			if err != nil {
+				if tc.ExpectSuccess {
+					t.Error(err)
+				} else { // FAILURE is expected
+					t.Logf("Create, expected result: %v", err)
+				}
+			}
+
+			apiEndPoints, err := clusterMgr.APIEndpoints()
+			if tc.ExpectSuccess {
+				if err != nil {
+					t.Error(err)
+				} else {
+					t.Logf("APIEndPoints %v", apiEndPoints)
+				}
+			} else { // FAILURE is expected
+				if err != nil {
+					t.Logf("APIEndpoints, expected result: %v", err)
+				}
+			}
+
+			err = clusterMgr.UpdateClusterStatus()
+			if tc.ExpectSuccess {
+				if err != nil {
+					t.Error(err)
+				}
+			} else { // FAILURE is expected
+				if err != nil {
+					t.Logf("UpdateClusterStatus, expected result: %v", err)
+				}
+			}
+			err = clusterMgr.Close()
+			if err != nil {
+				t.Error(err)
+			}
```

Comment:
This is testing other functions. Optimally, it would be split in their own tests for better readability, one test per function with a name <function>Test

Reply:
done